Synthetic data generation

almaan opened the pull request:
This PR contains some different updates - as discussed per Zulip with @giovp. To briefly summarize:

### datasets
- I've renamed the `pancreas.py` file to `sc_to_sp.py` since this could potentially hold more spatial datasets generated from single cell data. 
- the functions used to generate spatial data from a single cell dataset are all found in `datasets/_sc_to_sp_utils.py`. The `generate_synthetic_data` function is agnostic to which single cell data set it's presented with.
- Currently there are three datasets defined in the `sc_to_sp.py` all of them based on the `pancreas` dataset. The difference here is the dirichlet alpha parameter, which renders spatial observations with different degrees of cell type diversity/mixing of cell types.
- The pancreas single cell dataset is a combination of several different singe cell datasets, where some of these are processed (i.e., non-integers). The `sc_to_sp` process requires whole integer counts, hence we only use Smart-seq2 and InDrop derived cells in the pancreas data set. See : https://github.com/theislab/scib-reproducibility/tree/main/notebooks/data_preprocessing/pancreas

### anndata formatting
- After realizing that the anndata objects were written to disk during the test, causing an error to be thrown due to nested anndata files (we had `adata_spatial.uns["sc_reference"]` as an anndata object). We got [recommended](https://github.com/openproblems-bio/openproblems/pull/309#issuecomment-1103610776) to concatenate the single cell and spatial data into a single anndata object. We settled on this approach.
- two new methods have been added to the `spatial_decomposition/_utils.py``file:
  - `merge_sc_and_sp` : merges single cell and spatial data into a single object using `adata.concatenate`
  - `split_sc_and_sp` : splits the merged file and cleans up unnecessary features from the resulting two objects 
  
### methods update

giovp commented on openproblems/tasks/spatial_decomposition/_utils.py:
```diff
@@ -1,16 +1,83 @@
+    adata_sp = adata_merged[is_sp, :]
+    adata_sc = adata_merged[~is_sp, :]
+
+    # clean objects
```
again, unclear why we need to "delete" obs? Is it because a bunch of `np.nan` are added in concatenation? Not sure is huge problem in terms of storage. I'd approach it differentily by removing obs altogether and except for `cluster_label` of `adata_sc` ? but maybe again I'm missing something so feel free to disregard

almaan replied:
Yes exactly, I thought it was a bit unnecessary to have all these `np.nan` columns, but if you don't mind I'm fine with removing this part - I was honestly a bit hesitant to even put it in there!

giovp commented on openproblems/tasks/spatial_decomposition/datasets/_sc_to_sp_utils.py:
```diff
@@ -0,0 +1,169 @@
+
+    obs_names = ["spatial_{}".format(x) for x in range(n_obs)]
+    adata_spatial = ad.AnnData(
+        sp_x,
```
```suggestion
        csr_matrix(sp_x),
```

giovp replied:
can the count be sparse?

almaan replied:
actually, I initially had the matrix as sparse, but wasn't sure all the methods could handle it. Hence why I kept it as dense, but if you think it'll be fine that's great!

giovp commented on openproblems/tasks/spatial_decomposition/datasets/_sc_to_sp_utils.py:
```diff
@@ -0,0 +1,169 @@
+    rng = np.random.default_rng(seed)
+
+    # get single cell expression data
+    X = adata.X
```
```suggestion
    X = adata.X.A
```

giovp replied:
and this would be already dense

almaan replied:
ah neat, thanks!

giovp commented on openproblems/tasks/spatial_decomposition/datasets/_sc_to_sp_utils.py:
```diff
@@ -0,0 +1,169 @@
+        keep = keep | idx
+
+    adata = adata[keep, :]
+    adata.X = adata.X
```
what's this for?

almaan replied:
I honestly have no idea, this must be some accidental copy-pasting... I'll remove it!

giovp commented on openproblems/tasks/spatial_decomposition/datasets/_sc_to_sp_utils.py:
```diff
@@ -0,0 +1,169 @@
+        idx = adata.obs.tech.values == tech
+        keep = keep | idx
+
+    adata = adata[keep, :]
```
```suggestion
    adata = adata[keep, :].copy()
```

almaan replied:
thanks!

giovp commented on openproblems/tasks/spatial_decomposition/datasets/_sc_to_sp_utils.py:
```diff
@@ -0,0 +1,169 @@
+
+    adata_merged = merge_sc_and_sp(adata, adata_spatial)
+
+    return adata_merged
```
overall super nice and much cleaner refactoring 🚀

almaan replied:
Thank you! I bet it could have been done 100x better - but think it will do for now at least! Great feedback :pray:

giovp reviewed (changes requested):
minor comments

almaan commented on openproblems/tasks/spatial_decomposition/_utils.py:
```diff
@@ -1,16 +1,83 @@
+) -> ad.AnnData:
+
+    # merge single cell and spatial data
+    adata_merged = adata_sp.concatenate(
```
I didn't know the `adata.concatenate` was about to be depracated, makes full sense to use `ad.concat` instead!

almaan commented on openproblems/tasks/spatial_decomposition/_utils.py:
```diff
@@ -1,16 +1,83 @@
+    )
+
+    # get attributes from single cell and spatial data
+    sc_obs = set(adata_sc.obs.columns.tolist())
```
Maybe there's a better way of doing it, but I'm storing these in `uns` and use them in the `split_sc_and_sp` function to "clean up" the single cell and spatial anndata objects, removing unnecessary `.obs` or `.obsm` columns that are just filled with NaNs after the "outer joint".

almaan commented on openproblems/tasks/spatial_decomposition/_utils.py:
```diff
@@ -1,16 +1,83 @@
+def split_sc_and_sp(
+    adata_merged: ad.AnnData,
+    batch_key: str = "modality",
+    batch_categories: SpScBC = dict(spatial="sp", single_cell="sc"),
```
see answer above

almaan commented:
@giovp I've tried to act on all your comments (really awesome thanks!) from the review of the initial commit. The weird signature is removed, I've updated to `ad.concat`, removed the deletion single cell/spatial specific NaN `obs` and `obsm` in the merged object, and also added an import for `csr_matrix` (your change didn't include this).

Maybe one thing left to do is to make sure that all the methods are compatible with the sparse matrices?

giovp commented:
@almaan thanks a lot! looks great. I fixed the pancreas dataset which wasn't returning merged adata. Let's see if test pass now.

Must say one thing this (but probably other SCOP tasks) would really benefit are package constants utils. There's ton of hard-coded "keys" around the codebase that, if they'd have to be changed, would be a pain. No bandwidth for this atm so let's address it when the new platform is in place.

almaan commented:
> @almaan thanks a lot! looks great. I fixed the pancreas dataset which wasn't returning merged adata. Let's see if test pass now.
> 
> Must say one thing this (but probably other SCOP tasks) would really benefit are package constants utils. There's ton of hard-coded "keys" around the codebase that, if they'd have to be changed, would be a pain. No bandwidth for this atm so let's address it when the new platform is in place.

Awesome! Thanks for taking care of the pancreas data set!

I also agree that constants would prevent a lot of issues in the future, I wouldn't be keen on regexing all of the instants of certain keys or changing them manually... but also agree that there are more important stuff to focus on...

I've looked through failed tests and spotted two things that I believe are the culprits:

### `csr_matrix`

We get an error of the form `NotImplementedError: subtracting a nonzero scalar from a sparse matrix is not supported`  from `utils.asserts.assert_array_equal(adata.X, adata.layers["counts"])` in the core code's `test_load_dataset.py`. when comparing `adata.X` and `adata.layers["counts"]` (I think this is a part of the test suite). I managed to reproduce this error with:

```python
x = np.random.random((3,3))
np.testing.assert_array_equal(x,csr_matrix(x)) 
``` 

One quick and dirty fix is to not use sparse arrays in our synthetic data, but I'll see if there's a more elegant workaround.

###  modality `KeyError`

We get a `KeyError: 'modality'` error, when executing the `split_sc_and_sp` function. I haven't looked through this supermuch, but there must be some sort of typo or other thing going on here (ugh... constants _please_). I'll give it a browse and see if I can't fix this.

Hopefully things pass once these two points are taken care of.

giovp commented:
> One quick and dirty fix is to not use sparse arrays in our synthetic data, but I'll see if there's a more elegant workaround.

makes sense, let's just discard all sparse matrices.

> We get a KeyError: 'modality' error, when executing the split_sc_and_sp

yeah, for wahtever reason then nan are added (in the merge) the pandas series is not categorical anymore, will try a dirty fix today.

almaan commented:
> makes sense, let's just discard all sparse matrices.
>> On it!
> yeah, for wahtever reason then nan are added (in the merge) the pandas series is not categorical anymore, will try a dirty fix today.
>> I think I spotted the source of this (and revised my comment as you wrote your answer to the first draft - sorry..). WDYT? Does it make sense that the double merge could cause this?

almaan commented:
@giovp  I think I solved the nan problem, it wasn't the prettiest fix, but seems to have done the trick.

I tried to troubleshoot some of the stuff that failed after fixing the nans, and the errors all came when trying to install the `rgeos` package (I think this is a dependency of `sp` which is used by `Seurat`?). Looking into this a bit more, I think it's due to a system dependency (`geos`) missing (https://cran.r-project.org/web/packages/rgeos/index.html). Hence, I tried to add this to the r-extras Dockerfile according to: https://libgeos.org/usage/install/, but it keeps on failing... this time because the command to add an apt-repo from which `geos` can be installed breaks. More specifically, it's the command `add-apt-repository` that can't be found, even though I the added line: `RUN apt-get install python-software-properties` to the Dockerfile, which should take care of this (I think?)... I gotta do a bit more troubleshooting, but if you have any ideas please feel free to step in! Maybe @scottgigante-immunai has a good solution to this?

Also, apologies for the push-spam, but I'm currently working on a crap computer that can't handle Docker, so the only way to test a solution is to push...

scottgigante-immunai commented:
@almaan I think you're running `add-apt-repository` in both `r-base` and `r-extras`, but only installing it in `r-extras`. Also you might want to use `software-properties-common` rather than `python-software-properties`, and you should probably call `apt-get update` after `add-apt-repository`. Hopefully that helps.

scottgigante-immunai commented:
Update: the rgeos problem also hit on master, so I have installed `libgeos-dev` in https://github.com/openproblems-bio/openproblems/pull/380. Should be resolved once the latest builds complete in ~an hour.

giovp commented:
@scottgigante-immunai thanks a lot! I don't feel comfortable resolving conflicts for those lines, could you or @almaan do? then we can re-run and see if it's fixed on this branch

giovp commented:
@scottgigante-immunai I might have found the issue but had to pin anndata version >=0.8 , I did so in docker python-extra and r-extra, is that where it should be done? this issue arise in the data generation step (the `datasets` module) but couldn't trace back which docker that module uses.

giovp commented:
@scottgigante-immunai ok I see that anndata is pinned somewhere else and this is what I get from CI

`Successfully installed anndata-0.7.8 openproblems-0.6.0`

is it possible to upgrade to 0.8? also CI is failing for some other reasons I don't understand

scottgigante-immunai commented:
So the upgrade to anndata >=0.8 is a little tricky as the before and after are mutually incompatible and this would require dropping python 3.6 support. If there's any way we can fix the issue without upgrading anndata that would be much easier. If it's absolutely necessary I can try to get it working.

giovp commented:
ok I'll look into it.

giovp commented:
@scottgigante-immunai  I pinned anndata to 0.7.6 and it seems to work locally, I can load all datasets in correct shapes and annotations. Tests seem to still fail but I can't find any error message, do you have any idea/pointer on what could the problem be?
thank you

giovp commented:
I get this error while setting swap

```
Memory and swap:
               total        used        free      shared  buff/cache   available
Mem:           6.8Gi       771Mi       963Mi       9.0Mi       5.1Gi       5.7Gi
Swap:          4.0Gi        18Mi       4.0Gi

NAME          TYPE SIZE  USED PRIO
/mnt/swapfile file   4G 18.1M   -2

Run export SWAP_FILE=$(swapon --show=NAME | tail -n 1)
  export SWAP_FILE=$(swapon --show=NAME | tail -n 1)
  sudo swapoff $SWAP_FILE
  sudo rm $SWAP_FILE
  sudo fallocate -l 10G $SWAP_FILE
  sudo chmod 600 $SWAP_FILE
  sudo mkswap $SWAP_FILE
  sudo swapon $SWAP_FILE
  shell: bash --noprofile --norc -e -o pipefail {0}
  env:
    LINUX_VERSION: Linux 7511143b1ff7 5.13.0-1022-azure #26~[20](https://github.com/giovp/SingleCellOpenProblems/runs/6494704898?check_suite_focus=true#step:12:22).04.1-Ubuntu SMP Thu Apr 7 19:42:45 UTC 20[22](https://github.com/giovp/SingleCellOpenProblems/runs/6494704898?check_suite_focus=true#step:12:24) x86_64 GNU/Linux
    PYTHON_VERSION: Python 3.8.13
    R_VERSION: R version 4.2.0 (2022-04-22) -- "Vigorous Calisthenics"
swapoff: Not superuser.
```
I'll try to remove it and make smaller data.

scottgigante-immunai commented:
Opened an issue at https://github.com/pierotofy/set-swap-space/issues/2. They might be able to help.

giovp commented:
hi @scottgigante-immunai I don't know how to proceed here I think the data is already pretty small, can try to make it smaller though. What do you think?

scottgigante-immunai commented:
Your problem is [here](https://github.com/giovp/SingleCellOpenProblems/blob/synthetic-data-generation/openproblems/tasks/spatial_decomposition/datasets/_sc_to_sp_utils.py#L66), before you use `n_obs`.

giovp commented:
removed it but seems to still be a problem

scottgigante-immunai commented:
[This copy](https://github.com/giovp/SingleCellOpenProblems/pull/9/files#diff-d073766269bf40a8701083eda01027934a33080d6d89f263d516fb62546f125eR65) effectively doubles your memory use and it appears you don't use `X` at all?

giovp commented:
tests fails but much later and can't get log, could you re-run @scottgigante-immunai ? I commented out the failing tests from before.

scottgigante-immunai commented on .github/workflows/run_tests.yml:
```diff
@@ -182,6 +182,11 @@ jobs:
         snakemake -j $(nproc) docker
         cd ..
 
+    # - name: Set Swap Space
```
Best to remove this

scottgigante-immunai commented on docker/openproblems-python-extras/requirements.txt:
```diff
@@ -8,8 +8,10 @@ umap-learn
 git+https://github.com/michalk8/neuralee@master  # contains gradient error fix
 tensorflow-cpu==2.4.1
 git+https://github.com/czbiohub/molecular-cross-validation
+anndata>=0.7.6,<0.8
```
This should be set in `setup.py` since it's a requirement for all of the images, not just the extras

scottgigante-immunai commented on docker/openproblems-r-extras/requirements.txt:
```diff
@@ -1,2 +1,3 @@
 git+https://github.com/KrishnaswamyLab/harmonic-alignment#subdirectory=python
 git+https://github.com/theislab/scib
+anndata>=0.7.6,<0.8
```
This should be set in `setup.py` since it's a requirement for all of the images, not just the extras

scottgigante-immunai commented on openproblems/tasks/spatial_decomposition/README.md:
```diff
@@ -20,25 +20,17 @@ spatial data. We require that all datasets have an associated reference single
 
 
-## API
+## Datasets and API
```
Maybe leave this as API just for consistency

scottgigante-immunai commented on openproblems/tasks/spatial_decomposition/README.md:
```diff
@@ -20,25 +20,17 @@ spatial data. We require that all datasets have an associated reference single
-Metrics shall compare `adata.obsm['proportions_pred']` to `adata.obsm['proportions_true']`.
+In the single cell reference, cell-types are stored in `adata_sc.obs["label"]`.
+In the spatial target, ground-truth cell-type proportions are stored in `adata_spatial.obsm["proportions_true"]`.
+In the spatial target, ground-truth cell-type proportions are stored in `adata_spatial.obsm["proportions_true"]`.
```
Duplicated line

scottgigante-immunai commented on openproblems/tasks/spatial_decomposition/datasets/__init__.py:
```diff
@@ -1,2 +1,3 @@
-from .pancreas import pancreas_average
-from .pancreas import pancreas_cell
+# from .sc_to_sp import pancreas_alpha_0_1
```
Why are these commented out?

giovp replied:
I commented them out cause they were failing in CI and wanted to first check whether, when commented out, CI would pass.

scottgigante-immunai commented on openproblems/tasks/spatial_decomposition/datasets/_destvi_utils.py:
```diff
@@ -20,14 +23,44 @@
+
+
+@jit(nopython=True)
+def get_mean_normal(cell_types, gamma, mean_, components_):
```
Add `# pragma: no cover` to skip coverage on JIT compiled code. See https://github.com/openproblems-bio/openproblems/blob/main/openproblems/tasks/dimensionality_reduction/metrics/nn_ranking.py for an example

scottgigante-immunai commented on openproblems/tasks/spatial_decomposition/datasets/_destvi_utils.py:
```diff
@@ -181,20 +220,25 @@ def generate_synthetic_dataset(
-            plt.tight_layout()
-            # plt.savefig(output_dir+"lib.png")
-    return st_anndata
+        # if i == 0:
```
Remove commented code

scottgigante-immunai commented on openproblems/tasks/spatial_decomposition/datasets/_sc_to_sp_utils.py:
```diff
@@ -0,0 +1,170 @@
+
+    # get single cell expression data
+    X = adata.X
+    # if issparse(X):
```
Remove commented code

scottgigante-immunai commented on openproblems/tasks/spatial_decomposition/datasets/_sc_to_sp_utils.py:
```diff
@@ -0,0 +1,170 @@
+
+
+def get_pancreas_integer(adata: ad.AnnData):
+    # for some platforms the pancreas data set
```
This should probably be a docstring

scottgigante-immunai commented on openproblems/tasks/spatial_decomposition/datasets/sc_to_sp.py:
```diff
@@ -0,0 +1,33 @@
+from ._sc_to_sp_utils import generate_synthetic_dataset
+from ._sc_to_sp_utils import get_pancreas_integer
+
+# @dataset("Pancreas (alpha=1)")
```
Either uncomment or delete

scottgigante-immunai commented on openproblems/tasks/spatial_decomposition/datasets/sc_to_sp.py:
```diff
@@ -0,0 +1,33 @@
+    return merged_adata
+
+
+# @dataset("Pancreas (alpha=0.5)")
```
Either uncomment or delete

scottgigante-immunai replied:
If they are currently not working, remove from here and open a new draft PR that will add them in and we can fix it later

scottgigante-immunai commented on openproblems/tasks/spatial_decomposition/methods/rctd.R:
```diff
@@ -19,7 +19,7 @@ sce_sp <- sce[, colData(sce)$modality == "sp"]
 sp_counts <- assay(sce_sp, "X")
 # get spatial data coordinates
-sp_coords <- as.data.frame(sce_sp@reducedDims$spatial)
+sp_coords <- ad.data.frame(reducedDim(sce, "spatial"))
```
is `ad.data.frame` a typo? Why is this not causing failures?

scottgigante-immunai reviewed (changes requested):
I can't rerun the test since I don't have write access to your fork, but if you add another commit it will rerun

scottgigante-immunai commented:
I notice that is https://github.com/giovp/SingleCellOpenProblems/pull/9/files#diff-ac0e68b74d7a85d686733a3f9b7bf120bc6277fc3fea7e2828ae418c24319ae7R9 you pass `test=False` not `test=test`. Is that desired?

scottgigante-immunai commented:
It doesn't look like you use the `test` parameter at all.

scottgigante-immunai commented:
You might also try adding swap via docker args with `--memory-swap -1` in the [docker args](https://github.com/giovp/SingleCellOpenProblems/blob/master/.github/workflows/run_tests.yml#L132)

scottgigante-immunai commented:
And finally, if we continue running out of memory, we might need to move to self-hosted...

scottgigante-immunai commented:
Woohoo! We have new errors.

https://github.com/giovp/SingleCellOpenProblems/runs/6710574413?check_suite_focus=true#step:12:460 -- here you just have to copy `adata.__from_cache__` to your new AnnData object in the `utils._sp_to_sp` function.

The other one, looks like you need to call `FindVariableFeatures` somewhere.

giovp commented:
> https://github.com/giovp/SingleCellOpenProblems/runs/6710574413?check_suite_focus=true#step:12:460 -- here you just have to copy adata.__from_cache__ to your new AnnData object in the utils._sp_to_sp function.

@scottgigante-immunai where exactly should I copy the object? didn't get it
.

also, the other error comes from the batch_integration task which we didn't write so not really sure how to solve that
https://github.com/giovp/SingleCellOpenProblems/runs/6710574413?check_suite_focus=true#step:12:720

scottgigante-immunai commented:
@giovp you create a new AnnData [here](https://github.com/giovp/SingleCellOpenProblems/pull/9/files#diff-d073766269bf40a8701083eda01027934a33080d6d89f263d516fb62546f125eR147) -- so you need to set `adata_merged.__from_cache__ = adata.__from_cache__`

scottgigante-immunai commented:
The other issue is a bit weird. Your diff in this PR is very very hard to read so difficult for me to debug -- if you change the PR target to be `openproblems-bio/main` that should reduce the diff.

scottgigante-immunai commented:
(or alternatively merge `openproblems-bio/main` into `giovp/master` and then `giovp/master` into this branch.)

giovp commented:
> (or alternatively merge openproblems-bio/main into giovp/master and then giovp/master into this branch.)

done. I hope I didn't introduce errors in resolving the conflitcs

giovp commented:
I noticed a bunch of duplicates in the requirements files, is it intended?

https://github.com/giovp/SingleCellOpenProblems/blob/synthetic-data-generation/docker/openproblems-python-extras/requirements.txt

giovp commented:
tried to fix one error but the other one is strange
```
obj = AnnData object with n_obs × n_vars = 203 × 443
    obs: 'n_cells', 'tech', 'celltype', 'size_factors', 'n_counts', 'label', 'modality'
    obsm: 'n_cells', 'proportions_true', 'spatial'
    layers: 'counts', 'log_scran'
    def _py2rpy(obj):
        """ Dummy function for py2rpy.
        This function will convert Python objects into rpy2.rinterface
        objects.
        """
        if isinstance(obj, _rinterface_capi.SupportsSEXP):
            return obj
>       raise NotImplementedError(
            "Conversion 'py2rpy' not defined for objects of type '%s'" %
            str(type(obj))
        )
E       NotImplementedError: Conversion 'py2rpy' not defined for objects of type '<class 'anndata._core.anndata.AnnData'>'
```

giovp commented:
@scottgigante-immunai ok now the two errors are:

- findVariableFeatures from the batch integration task
- rpy2 conversion in the spatial task, which I unfortunately don't know how to address

scottgigante-immunai commented on openproblems/tasks/spatial_decomposition/_utils.py:
```diff
@@ -1,16 +1,61 @@
+            adata_sc.obsm[k] = df
+
+    # merge single cell and spatial data
+    adata_merged = ad.concat(
```
Maybe try using `ad.concatenate` since `ad.concat` is still experimental in 0.7.x

giovp commented:
> My suspicion is that the batch integration error comes from pinning Seurat==4.0.1. Is that necessary?

I don't think it is @almaan do you know? I believe it wasn't set by us?

giovp commented on docker/openproblems-python-extras/requirements.txt:
```diff
@@ -1,11 +1,14 @@
 git+https://github.com/atong01/SCOT
 git+https://github.com/czbiohub/molecular-cross-validation
 git+https://github.com/czbiohub/molecular-cross-validation
+git+https://github.com/czbiohub/molecular-cross-validation
```
will do but btw not sure why they were introduced...

giovp commented:
logging out, will work on it tomorrow!

giovp commented on openproblems/tasks/spatial_decomposition/datasets/sc_to_sp.py:
```diff
@@ -0,0 +1,39 @@
+    adata.obs["label"] = adata.obs["celltype"]
+
+    merged_adata = generate_synthetic_dataset(adata, n_obs=n_obs, alpha=1)
+    merged_adata.__from_cache__ = from_cache
```
I didn't know what this was and copied it from some other place in the repo. WIll change accordingly.

scottgigante-immunai commented on docker/openproblems-python-extras/requirements.txt:
```diff
@@ -1,25 +1,15 @@
 cmake==3.18.4
 git+https://github.com/atong01/SCOT
 git+https://github.com/czbiohub/molecular-cross-validation
-git+https://github.com/czbiohub/molecular-cross-validation
```
This diff should be fixed on `master` to make this easier to read...

scottgigante-immunai commented on openproblems/tasks/spatial_decomposition/datasets/_sc_to_sp_utils.py:
```diff
@@ -0,0 +1,173 @@
+        idx = adata.obs.tech.values == tech
+        keep = keep | idx
+
+    adata = adata[keep, :].copy()
```
Looks like this line might also clear `adata.__from_cache__`

scottgigante-immunai commented on openproblems/tasks/spatial_decomposition/methods/destvi.py:
```diff
@@ -1,5 +1,6 @@
 from ....tools.decorators import method
 from ....tools.utils import check_version
+from .._utils import split_sc_and_sp
```
If we're importing methods from `_utils`, it shouldn't be a private submodule -- rename to `utils`

LuckyMD commented:
@giovp @almaan could we merge this PR even if things are not perfect yet and then edit the other PR to the main repo ([#309](https://github.com/openproblems-bio/openproblems/pull/309)). @scottgigante-immunai has edit access in the other PR so could fix the issues quickly there.

Also, we should document all things that need fixing and maybe add them to the FAQ for future troubleshooting.
